api/src/opentrons/protocol_engine/state/_labware_origin_math.py

```
@@ -0,0 +1,236 @@
+
+        module_offset_point = _to_point_from_lw_offset_vector(
+            module_parent_to_child_offset
+        )
+        return module_offset_point - child_labware_overlap_with_parent_entity
+
+    elif isinstance(labware_location, OnLabwareLocation):
+        assert isinstance(parent_entity, (LabwareDefinition2, LabwareDefinition3))
+
+        # TODO(jh, 06-05-25): This logic is slightly duplicative of LabwareView get_dimensions. Can we unify?
+        if isinstance(parent_entity, LabwareDefinition2):
+            parent_entity_height = parent_entity.dimensions.zDimension
+        else:
+            assert_type(parent_entity, LabwareDefinition3)
+            parent_entity_height = (
+                parent_entity.extents.total.frontRightTop.z
+                - parent_entity.extents.total.backLeftBottom.z
+            )
+
+        child_labware_overlap_with_parent_entity = (
+            _get_child_labware_overlap_with_parent_labware(
+                child_labware=child_labware,
+                parent_labware_name=parent_entity.parameters.loadName,
+            )
+        )
+
+        return Point(
+            x=child_labware_overlap_with_parent_entity.x,
+            y=child_labware_overlap_with_parent_entity.y,
+            z=parent_entity_height - child_labware_overlap_with_parent_entity.z,
+        )
+
+    else:
+        raise TypeError(f"Unsupported labware location type: {labware_location}")
+
+
+def _get_child_labware_overlap_with_parent_labware(
+    child_labware: LabwareDefinition, parent_labware_name: str
+) -> Point:
+    """Get the child labware's overlap with the parent labware's load name."""
+    if parent_labware_name in child_labware.stackingOffsetWithLabware.keys():
+        child_labware_overlap = child_labware.stackingOffsetWithLabware[
+            parent_labware_name
+        ]
+    else:
+        child_labware_overlap = child_labware.stackingOffsetWithLabware.get(
+            "default", Vector3D(x=0, y=0, z=0)
+        )
```

I'm a very slightly suspicious of this final `stackingOffsetWithLabware` fallback to (0, 0, 0), but let me know what you think.

Our code to determine whether stacking is allowed was getting uncomfortably scattered. [We're reining it in a bit.](https://github.com/Opentrons/opentrons/pull/18460#discussion_r2116237277) 

So I'm wondering:

* Instead of falling back to (0, 0, 0), should this raise some kind of "stack not allowed" error?
* Should we try to deduplicate the logic ("the logic" being searching through `stackingOffsetWithLabware`, falling back to `default`, possibly taking other things into account like whether the labware is a `lid`, ...) between here and `raise_if_labware_cannot_be_stacked()`?

No action needed necessarily, just exploring what we want the boundaries of this module to be.

---

Thanks for the context. Yeah, I think I'm down to throw an error here. Seems sensible.